test/framework/src/main/java/org/elasticsearch/test/TestCluster.java

```diff
@@ -69,14 +71,111 @@ public void beforeTest(Random randomGenerator) throws IOException, InterruptedEx
      * Wipes any data that a test can leave behind: indices, templates (except exclude templates) and repositories
      */
     public void wipe(Set<String> excludeTemplates) {
-        // First delete data streams, because composable index templates can't be deleted if these templates are still used by data streams.
-        wipeAllDataStreams();
-        wipeAllComposableIndexTemplates(excludeTemplates);
-        wipeAllComponentTemplates(excludeTemplates);
-
-        wipeIndices("_all");
-        wipeAllTemplates(excludeTemplates);
-        wipeRepositories();
+        if (size() == 0) {
+            return;
+        }
+        final PlainActionFuture<Void> done = new PlainActionFuture<>();
+        try (RefCountingListener refCountingListener = new RefCountingListener(done)) {
+            wipeAllTemplates(excludeTemplates, refCountingListener.acquire());
+            // First delete data streams, because composable index templates can't be deleted if these templates are still used by data
+            // streams.
+            client().execute(
+                DeleteDataStreamAction.INSTANCE,
+                new DeleteDataStreamAction.Request(ESTestCase.TEST_REQUEST_TIMEOUT, "*").indicesOptions(
+                    IndicesOptions.LENIENT_EXPAND_OPEN_CLOSED_HIDDEN
+                ),
+                new DelegatingActionListener<>(refCountingListener.acquire()) {
+                    @Override
+                    public void onResponse(AcknowledgedResponse acknowledgedResponse) {
+                        assertAcked(acknowledgedResponse);
+                        wipeIndicesAsync(
+                            new String[] { "_all" },
+                            refCountingListener.acquire().delegateFailure((l, r) -> wipeRepositories(l))
+                        );
+                        deleteTemplates(excludeTemplates, delegate);
+                    }
+
+                    @Override
+                    public void onFailure(Exception e) {
+                        // Ignore if action isn't registered, because data streams is a module and
+                        // if the delete action isn't registered then there no data streams to delete.
+                        if (e.getMessage().startsWith("failed to find action") == false) {
+                            delegate.onFailure(e);
+                        } else {
+                            onResponse(AcknowledgedResponse.TRUE);
+                        }
+                    }
+                }
+            );
+        }
+        try {
+            done.get();
```

Comment:
Likewise here, can we bound the wait and handle interrupts properly?

Reply:
Jup, as with the other one, just went for `safeGet` now, 10s have to suffice IMO.